File: elixir/bob/bob.exs
defmodule Teenager do
  def hey(input) do
    cond do
      Regex.match? ~r/^ *$/, input ->
        "Fine. Be that way!"

      (! Regex.match? ~r/^[\d\W]+$/, input) and
      String.upcase(input) == input ->
        "Whoa, chill out!"

      String.last(input) == "?" ->
        "Sure."

      true ->
        "Whatever."
    end
  end
end

File: haskell/bob/src/Bob.hs
module Bob
  ( responseFor
  ) where

import Data.Char (isUpper, isAlpha, isSpace)

responseFor :: String -> String
responseFor xs
  | all isSpace xs = "Fine. Be that way!"
  | any isAlpha xs && all isUpperIfAlpha xs = "Whoa, chill out!"
  | last' (filter (not . isSpace) xs) == Just '?' = "Sure."
  | otherwise = "Whatever."
  where
    isUpperIfAlpha c = isUpper c || not (isAlpha c)
    last' [] = Nothing
    last' xs = Just (last xs)
